src/main/java/org/folio/rest/impl/TenantPermsAPI.java

```diff
@@ -35,46 +42,386 @@
           PostTenantpermissionsResponse.respond500WithTextPlain(e.getMessage())));
     }
   }
 
-  private Future<Void> savePermList(List<Perm> permList, Context vertxContext, String tenantId) {
+  private Future<Permission> addMissingModuleContext(Permission perm, ModuleId moduleId,
+      Context vertxContext, String tenantId) {
+    Promise<Permission> promise = Promise.promise();
+    perm.setModuleName(moduleId.getProduct());
+    SemVer semver = moduleId.getSemVer();
+    perm.setModuleVersion(semver != null ? semver.toString() : null);
+
+    PostgresClient.getInstance(vertxContext.owner(), tenantId).update(TABLE_NAME_PERMS, perm,
+        perm.getId(), updateReply -> {
+          if (updateReply.failed()) {
+            Throwable cause = updateReply.cause();
+            logger.error(cause.getLocalizedMessage(), cause);
+            promise.fail(cause);
+            return;
+          }
+          logger.info(String.format("Added defined context to %s", perm.getPermissionName()));
+          promise.complete(perm);
+        });
+    return promise.future();
+  }
+
+  private Future<List<Permission>> getPermsForModule(ModuleId moduleId, List<OkapiPermission> perms,
+      Context vertxContext, String tenantId) {
+    return getPermsByModule(moduleId, vertxContext, tenantId)
+        .compose(existing -> {
+          if (existing.isEmpty() && perms != null) {
+            // this means either:
+            // A) the first time enabling this module, or
+            // B) the permissions exist but don't yet have the moduleName field
+            List<Permission> ret = new ArrayList<>();
+            List<Future> futures = new ArrayList<>(perms.size());
+            perms.forEach(perm ->
+              futures.add(getModulePermByName(perm.getPermissionName(), null, vertxContext, tenantId)
+                  .compose(dbPerm -> {
+                    if (dbPerm != null && dbPerm.getModuleName() != null
+                        && Boolean.FALSE.equals(dbPerm.getDummy())) {
+                      // Edge case of (A) where another module already has a permission with
+                      // the same name. We need to fail here as there isn't anything we can do
+                      String msg = String.format(
+                          "Collision! A Permission with name %s is already defined by %s",
+                          perm.getPermissionName(), dbPerm.getModuleName());
+                      logger.error(msg);
+                      return Future.failedFuture(msg);
+                    }
+                    if (dbPerm != null && PermissionUtils.equals(perm, dbPerm)) {
+                      // (B) we have a match, but lack moduleName. Fix it before we add it.
+                      return addMissingModuleContext(dbPerm, moduleId, vertxContext, tenantId)
+                          .onFailure(Future::failedFuture)
+                          .onSuccess(ret::add);
+                    }
+                    return Future.succeededFuture();
```

Comment:
Would code block below do what we want?
```
                    if (dbPerm == null) {
                      return Future.succeededFuture();
                    }
                    if (dbPerm.getModuleName() == null && !dbPerm.getMutable() 
                        && PermissionUtils.equals(perm, dbPerm)) {
                      // (B) we have a match, but lack moduleName. Fix it before we add it.
                      return addMissingModuleContext(dbPerm, moduleId, vertxContext, tenantId)
                          .onFailure(Future::failedFuture)
                          .onSuccess(ret::add);
                    } else {
                      String msg = String.format(
                          "Collision! A Permission with name %s is already defined by %s",
                          perm.getPermissionName(), Json.encode(dbPerm));
                      logger.error(msg);
                      return Future.failedFuture(msg);
                    }
```

Reply:
No.  Like I said earlier we need to check dummy. Otherwise we break how dummy permissions are used as placeholders for subPermissions defined by other modules until the real thing is defined.  This is exercised in the unit tests.  See RestVerticleTest, specifically relevant are tests involving "sendOtherPermissionSet" and "sendAlienPermissionSet".

Also, sonarqube will complain about not testing mutable against the primitive on that mutable check, hence the uglier (but safer) syntax is preferred.  Possible NPE otherwise.

I will add the bit about case B matching on mutable too though.

Reply:
I did not mean to fix your syntax. I was just trying to help to get the logic right. For example, current implementation does not handle the case where the found duplicated perm name is a user defined perm, or dummy perm. BTW dummy perm is just the precursor of normal perm.

Reply:
Ah, got it @hjiebsco.  I was looking at this/responding on my phone last night and it appeared that this was a code suggestion to be applied/committed as-is.  

> For example, current implementation does not handle the cases where the found duplicated perm name is a user defined perm, or dummy perm

The cases we miss here are caught later in the processing, but it's probably best to fail early.  

> dummy perm is just the precursor of normal perm

Exactly.  That's what I was trying to say, which is why we can't consider it a collision if the found permission is a dummy.  

I've taken most your suggestion, made the necessary syntax adjustments and also continue w/o failing if the found permission is a dummy.  I think this covers all the bases.

Reply:
> Ah, got it @hjiebsco. I was looking at this/responding on my phone last night and it appeared that this was a code suggestion to be applied/committed as-is.
> 
> > For example, current implementation does not handle the cases where the found duplicated perm name is a user defined perm, or dummy perm
> 
> The cases we miss here are caught later in the processing, but it's probably best to fail early.
> 
> > dummy perm is just the precursor of normal perm
> 
> Exactly. That's what I was trying to say, which is why we can't consider it a collision if the found permission is a dummy.
> 
> I've taken most your suggestion, made the necessary syntax adjustments and also continue w/o failing if the found permission is a dummy. I think this covers all the bases.

About the dummy perm, for example if mod-A uses a perm but did not define it, it will become a dummy leftover. BTW mod-perm should not allow that but that's another defect for another day. Later when mod-B comes in and it also defines the same perm. This is a collision in my view, isn't it?

Reply:
Not in my mind.  If it's a dummy permission it's just a placeholder and can safely be replaced with a non-dummy permission.

Reply:
Dummy is just a concept in mod-permissions' world. From developers' perspective, they have no idea about whether their perms will be tagged as dummy or not. In any case, duplicated permission names between modules is a collision, isn't? Anyway, this would not be a problem if mod-perms does not have that pre-existing defect.

Reply:
We discussed this off-line and decided to take this to the rest of the team for discussion and possibly follow-on work.  We agreed it shouldn't block this PR from being merged.